### PR TITLE
osd/PG.cc: if osd has osdmap gap,find_best_info need checkout one more time whether this osd can be the authority

### PR DESCRIPTION
**Pull request by mingyuez**
fix osdmap gap problem

http://lists.ceph.com/pipermail/ceph-users-ceph.com/2017-July/019395.html

Signed-off-by: mingyuez <zhao.mingyue@h3c.com>

### CODE REVIEW

File: src/osd/PG.cc [hunk 1]
```diff
@@ -311,7 +311,14 @@ void PG::proc_master_log(
...
     info.last_epoch_started = oinfo.last_epoch_started;
-  info.history.merge(oinfo.history);
+  if (info.history.merge(oinfo.history)) {
+    if (info.history.last_epoch_clean >= info.history.same_interval_since) {
+      dout(20) << __func__ << "clearing past_intervals" << dendl;
+      past_intervals.clear();
+      dirty_big_info = true;
+    }
+  }
+
```
Comment (liu-chunmei) [inline src/osd/PG.cc]: in the peering process, in what situation last_epoch_clean will be >= same_interval_since? last_epoch_clean  is the epoch of last finishing recovery, same_interval_since is the epoch at this peering, right?

Comment (mingyuez) [inline src/osd/PG.cc]: the same_interval_since is the start of one past_interval,  and info.history.last_epoch_clean is the epoch when pg chage to active+clean(PG::mark_clean can update,and replica received primary msg may be update);
In normal condition, last_epoch_clean can be bigger than same_interval_since;

File: src/osd/PG.cc [hunk 2]
```diff
@@ -362,8 +369,15 @@ bool PG::proc_replica_info(
...
-  if (info.history.merge(oinfo.history))
+  if (info.history.merge(oinfo.history)) {
     dirty_info = true;
+    if (info.history.last_epoch_clean >= info.history.same_interval_since) {
+      dout(20) << __func__ << "clearing past_intervals" << dendl;
+      past_intervals.clear();
+      dirty_big_info = true;
+    }
+  }
+
```
Comment (liu-chunmei) [inline src/osd/PG.cc]: this kind of code seems already exist in update_history and called by proc_master_log and proc_replica_info

File: src/osd/PG.cc [hunk 3]
```diff
@@ -4833,7 +4870,12 @@ void PG::start_peering_interval(
     dout(10) << __func__ << ": check_new_interval output: "
 	     << debug.str() << dendl;
     if (new_interval) {
-      dout(10) << " noting past " << past_intervals.rbegin()->second << dendl;
+      if ((osdmap->get_epoch() == osd->get_superblock().oldest_map) && (info.history.last_epoch_clean < osdmap->get_epoch())) {
+        dout(5) << "map gap, cleaning past_intervals and faking" << dendl;
+        past_intervals.clear();
+      } else {
+        dout(10) << " noting past " << past_intervals.rbegin()->second << dendl;
+      }
```
Comment (liu-chunmei) [inline src/osd/PG.cc]: please check the latest version, this code seems exist there.

Comment (mingyuez) [inline src/osd/PG.cc]: hmm,I know clear past_intervals has fixed in luminous, but osdmap gap problem did not solve ; this PR is on hammer version, and I find that fix osdmap gap  need clear past_intervals,so I backport it

**Comment by smithfarm** [comment]
Hammer is EOL.

**Pull request closed without merging**